spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/support/DefaultSubscriberFactory.java

```diff
@@ -282,4 +321,85 @@ public SubscriberStub createSubscriberStub() {
 		}
 	}
 
+	/**
+	 * Creates {@link ExecutorProvider}. If a custom executor provider is set then the
+	 * subscriber properties configured through the application.properties file will be
+	 * ignored.
+	 * @param subscriptionName subscription name
+	 * @return executor provider
+	 */
+	public ExecutorProvider getExecutorProvider(String subscriptionName) {
+		if (this.executorProvider != null) {
+			return this.executorProvider;
+		}
+		if (this.pubSubConfiguration != null) {
+			PubSubConfiguration.Subscriber subscriber = this.pubSubConfiguration.getSubscriber(subscriptionName,
+					this.projectId);
+			return getExecutorProviderFromConfigurations(subscriber, subscriptionName);
+		}
+		return null;
+	}
+
+	/**
+	 * Creates {@link ExecutorProvider} for subscriber when Pub/Sub configurations (either
+	 * global/default or subscription-specific) are present.
+	 * @param subscriber subscriber properties
+	 * @param subscriptionName subscription name
+	 * @return executor provider
+	 */
+	ExecutorProvider getExecutorProviderFromConfigurations(PubSubConfiguration.Subscriber subscriber,
+			String subscriptionName) {
+		if (!subscriber.isGlobal()) {
+			return createExecutorProvider(subscriber, subscriptionName);
+		}
+
+		if (this.defaultExecutorProvider != null) {
+			return this.defaultExecutorProvider;
+		}
+		this.defaultExecutorProvider = createExecutorProvider(subscriber, subscriptionName);
+		return this.defaultExecutorProvider;
+	}
+
+	ExecutorProvider createExecutorProvider(PubSubConfiguration.Subscriber subscriber, String subscriptionName) {
+		if (this.executorProviderMap.containsKey(subscriptionName)) {
+			return this.executorProviderMap.get(subscriptionName);
+		}
+		ThreadPoolTaskScheduler scheduler = createThreadPoolTaskScheduler(subscriber, subscriptionName);
+		scheduler.initialize();
+		ExecutorProvider executor = FixedExecutorProvider.create(scheduler.getScheduledExecutor());
+		return this.executorProviderMap.computeIfAbsent(subscriptionName, k -> executor);
+	}
+
+	/**
+	 * Creates {@link ThreadPoolTaskScheduler} given subscriber properties.
+	 * @param subscriptionName subscription name
+	 * @return thread pool scheduler
+	 */
+	ThreadPoolTaskScheduler createThreadPoolTaskScheduler(PubSubConfiguration.Subscriber subscriber,
+			String subscriptionName) {
+		if (this.threadPoolTaskSchedulerMap.containsKey(subscriptionName)) {
+			return threadPoolTaskSchedulerMap.get(subscriptionName);
+		}
+		ThreadPoolTaskScheduler scheduler = new ThreadPoolTaskScheduler();
```

Comment:
Hey @elefeint @meltsufin @mpeddada1 !
By direct instantiation of `ThreadPoolTaskScheduler` here I think we lose the ability of distributed tracing instrumentation via [ExecutorBeanPostProcessor](https://github.com/spring-cloud/spring-cloud-sleuth/blob/main/spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/instrument/async/ExecutorBeanPostProcessor.java).
It should be exposed as a bean (as it was before the MR) or [explicitly wrapped in a dedicated class](https://docs.spring.io/spring-cloud-sleuth/docs/current/reference/html/integrations.html#sleuth-async-executor-service-integration)

Reply:
I guess the problem here is that there would potentially be a separate instance of such a bean for each of the subscriptions. @mpeddada1 Can you look into how to keep these beans in the application context so that they can be exposed to `ExecutorBeanPostProcessor`?

Reply:
I see. Created an issue to track this, thank you.